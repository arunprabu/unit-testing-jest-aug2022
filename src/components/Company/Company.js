import React, { useState } from 'react'

const Company = (props) => {

  const [myCountry, setMyCountry] = useState('');

  const handleCountryNameChange = (event) => {
    setMyCountry(event.target.value);
  }

  return (
    <div>
      <h2>Company | Testing Props, Events, States, Hooks, JSX and Snapshot</h2>
      <h3>Welcome to My Company's official Website!</h3>
      <div data-testid="companyNameEl">
        {props.name}
      </div>
      
      <p>Employee Count:</p>
      <p data-testid="employeeCountEl"
       style={{color: '#0000ff'}}>318000+</p>

      <input type="text"
       placeholder='Enter Country' value={myCountry} 
       onChange={ handleCountryNameChange }/>

    </div>
  )
}

export default Company;