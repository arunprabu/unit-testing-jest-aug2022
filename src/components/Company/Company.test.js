import { render, screen, fireEvent } from "@testing-library/react";
import Company from "./Company";
import renderer from 'react-test-renderer';

describe('CompanyComponent', () => { 

  
  // setting up
  // beforeEach( () => {
  // })

  // beforeAll( () => {

  // });

  // Tear down 
  // afterEach
  // afterAll 

  // Test spec #1
  it('should be available within app', ()=> {
    expect(Company).toBeDefined();
  });

  // Test Spec #2 - Let's test the props
  it('receives props and displays them properly', () =>{
    // rendering the comp with props 
    render(<Company name='Cognizant Technologies Pvt Ltd'/>);
    // act 
    // fetching the element 
    const companyNameEl = screen.getByTestId('companyNameEl');
    
    // assert 
    // checking if the prop is displayed inside the element or not
    expect(companyNameEl).toHaveTextContent('Cognizant Technologies Pvt Ltd');
  });

  // Test spec #3
  it('has Welcome to My Company\'s official Website!', () => {
    render(<Company />);
    const officialTxtEl = screen.getByText(`Welcome to My Company's official Website!`);
    expect(officialTxtEl).toBeInTheDocument();
  });

  // Test Spec #4
  it('has text with css color green', () => {
    render(<Company />);
    // expect(screen.getByTestId('employeeCountEl'))
    // .toHaveStyle('color: green');

    // if you are working with color code -- try the following 
    // use  rgb code or hexadecimal
    expect(screen.getByTestId('employeeCountEl')).toHaveStyle('color: #0000ff');
  });


  // Test spec #5
  // TODO: test whether the el has class  'btn'  
  // my requirement: Do Not use getByTestId() . instead use 'getByRole()
  // Have the button with label being 'JOIN COGNIZANT' 

  // Test Spec #6
  it(`has an input element with placeholder 'Enter Country'`, () => {
    render(<Company />);
    expect(screen.getByPlaceholderText('Enter Country')).toBeTruthy();
  });

  // Test Spec #7. Testing events, useState hook
  it(`should update country on onChange event`, () => {
    render(<Company />);
    // find the element with placeholder 'Enter Country' 
    const countryInput = screen.getByPlaceholderText('Enter Country');
    // now checking if onChange is working or not 
    // mock fire change event with the mock value -- all thru program
    fireEvent.change(countryInput, { target: {value: 'Canada'}});
    // finally checking if the value of input element is showing the entered country or not
    expect(countryInput.value).toBe('Canada');
  });

  // Snapshot Testing
  it('should have right company comp snapshot', () => {
    // to take snapshot -- we need a tool named react-test-renderer (npm i react-test-renderer )
    // taking a snapshot and convert it into JSON
    // [RECOMMENDED]: Take snapshots with necessary props
    const snapshotJSON = renderer.create(<Company name='Cognizant'/>).toJSON();
    // let's assert with a matcher toMatchSnapshot() from jest
    expect(snapshotJSON).toMatchSnapshot();
  }); 

  
})