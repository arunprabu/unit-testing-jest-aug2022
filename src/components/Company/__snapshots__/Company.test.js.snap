// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyComponent should have right company comp snapshot 1`] = `
<div>
  <h2>
    Company | Testing Props, Events, States, Hooks, JSX and Snapshot!!!
  </h2>
  <h3>
    Welcome to My Company's official Website!
  </h3>
  <div
    data-testid="companyNameEl"
  >
    Cognizant
  </div>
  <p>
    Employee Count:
  </p>
  <p
    data-testid="employeeCountEl"
    style={
      Object {
        "color": "#0000ff",
      }
    }
  >
    318000+
  </p>
  <input
    onChange={[Function]}
    placeholder="Enter Country"
    type="text"
    value=""
  />
</div>
`;
